### ISSUE TITLE
PropertyChanged with propertyName="" is not refreshing binded properties

### ISSUE THREAD

**Issue by dimitarcnbs** [xamarin/Xamarin.Forms#5121]
### Description
When PropertyChanged is called with propertyName="" ALL properties are considered as changed. 
That should trigger all binded properties on the XAML Page to refresh.

### Expected Behavior
1. this.PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(""));
2. Binded XMAL properties values should change

### Actual Behavior
1. this.PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(""));
2. Binded XMAL properties values did not change

### Basic Information

- Version with issue: 4.0.0.8055-pre1, 4.0.0.8055-pre2, 4.0.0.8055-pre3
- Last known good version: 3.4.0.1029999
- Platform Target Frameworks:
  - iOS:  12.1

**Comment by andreinitescu** [comment]
I took some time to test this with 4.0.0.8055-pre1 and latest 4.0.0.94569-pre3 on on iOS 12.1, and it works fine 

> Version with issue: 4.0.0.8055-pre1, 4.0.0.8055-pre2, 4.0.0.8055-pre3

These versions don't exist: 4.0.0.8055-pre2, 4.0.0.8055-pre3

@dimitarcnbs  While debugging the app, does Output tab show any binding errors?

**Comment by StephaneDelcroix** [comment]
We have tests for that, and it should work.

As @andreinitescu said: please provide a reproduction project.

**Comment by StephaneDelcroix** [comment]
please reopen as soon as you have a repro. closing for now.

**Comment by dimitarcnbs** [comment]
@samhouts could you provide me with the location of the UI test for this feature? I would like to use it as a starting point to attempt to build a repro. There is a braking change between the version I'm using and the latest version of Xamarin.Forms that affects my project. I need to find and resolve this issue before I could move forward and use the latest Xamarin.Forms version.

**Comment by dimitarcnbs** [comment]
By looking at the code I may have found where the issue I'm observing is rooted. 
I'm only looking at the code, and I have not been able to execute or debug it, and my interpretation may be faulty... 

In \Xamarin.Forms\Xamarin.Forms.Core\BindingExpression.cs
```
void ApplyCore(object sourceObject, BindableObject target, BindableProperty property, bool fromTarget = false)
```

part of the code has changed to:

	if (part.NextPart != null &&   (mode == BindingMode.OneWay || mode == BindingMode.TwoWay)
	    && current is INotifyPropertyChanged inpc)
			part.Subscribe(inpc);

It looks like for binding expression where part.IsSelf is true and part.NextPart is null a.k.a. {Binding .}  part.Subscribe(inpc); would never execute. Therefore changes to sourceObject would not be intercepted. 

This is a change in the code that altered its behavior. The code before the change used to execute part.Subscribe(inpc); in this case.

**Comment by dimitarcnbs** [comment]
@samhouts is it possible to reopen this issue based on the input I provided here, or I need to create new one?

**Comment by StephaneDelcroix** [comment]
@dimitarcnbs 
> please reopen as soon as you have a repro

this is still valid :) come up with a simple case (one xaml file, one view model) and we will reopen. Thanks

**Comment by dimitarcnbs** [comment]
@StephaneDelcroix 
[App5.zip](https://github.com/xamarin/Xamarin.Forms/files/3184164/App5.zip)
Attached is a repro. It has two labels and a button (MainPage.xaml, MainPage.xaml.cs).
When button is pushed the labels change value and the values should be the same. With Xamarin.Forms 3.4 it works. With 3.6 it doesn't work.

**Comment by StephaneDelcroix** [comment]
this is the unites we currently have for this: https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Core.UnitTests/BindingUnitTests.cs#L1952-L1968

**Comment by StephaneDelcroix** [comment]
Thanks for providing a sample for the reported issue. As indicated in the previous comment, we indeed had a test case for this.

I think the current behavior is correct, and the previous one was buggy. The doc says (https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.inotifypropertychanged.propertychanged?redirectedfrom=MSDN&view=netframework-4.8):

> ...The PropertyChanged event can indicate all properties on the object have changed...

And that means all properties of the current object (`Prop1`) in your case, but not the current object itself (`.`). If you want that behavior, you should generate an event indicating that the BindingContext for the page has changed.

@samhouts ok to close this again, for good ?

**Comment by dimitarcnbs** [comment]
You are correct, the current object did not change - the instance is the same, however all of its properties changed. Now, the Convertor that is taking this object and returns a value compiled of its (all/many) properties needs to be triggered when the object's properties change to return a value. There is really not a good way to write converters that are based on all of the object's properties. 

The case logic goes like that - I can have Converter triggered when Property1 changes, when Property2 changes, or when all properties changed - oh wait is there a trigger that is triggered when all properties changed - oh yes, and how to tap into that, well "." is the only option we have as of now.
I believe the "." behavior I describe is how the windows xaml works (or worked 2years ago).

Please think about covering this scenario in the xaml bindings. The previous implementation did cover it. It's very hard and not trivial to write a work around since the implementation change.

**Comment by samhouts** [comment]
[Issue5121_WPF.zip](https://github.com/xamarin/Xamarin.Forms/files/3341094/Issue5121_WPF.zip)

@dimitarcnbs I've attached a WPF project using basically the same code as your repro project, and it behaves exactly the same. We strive to have the same behavior as WPF/UWP XAML, so I believe that this is expected behavior. 

For this type of scenario, I would typically expose a calculated property on the ViewModel and raise its changed event when one of its dependent properties changed.

There is always room for us to improve. I can classify this as an enhancement, and we can see if we have added interest in it.

Thanks!

**Comment by dimitarcnbs** [comment]
@samhouts Thank you! 
I can see your point of view and I agree with the current implementation. At that point something like "multi-binding" is needed to achieve the functionality I described. There are solutions out there and I'm using one of them, however they are not ideal. 
A "build-in" implementation provided with Xamarin.Forms would be greatly appreciated.

**Comment by jfversluis** [comment]
I think this is fixed for .NET MAUI! 

Thanks for this suggestion! As Xamarin.Forms is now in maintenance mode, this will not happen anymore for Xamarin.Forms. We're only adding bugfixes and stability fixes.

If this is still important to you, make sure to check the [.NET MAUI repo](https://github.com/dotnet/maui) and see if it's already on the roadmap. If not, feel free to open a discussion to discuss a change first or open an issue with a detailed feature request. Thanks!
